src/Compilers/CSharp/Portable/Emitter/Model/TypeParameterSymbolAdapter.cs

```diff
@@ -223,33 +223,50 @@ Cci.ITypeReference Cci.IGenericTypeParameterReference.DefiningType
         IEnumerable<Cci.TypeReferenceWithAttributes> Cci.IGenericParameter.GetConstraints(EmitContext context)
         {
             var moduleBeingBuilt = (PEModuleBuilder)context.Module;
-            var seenValueType = false;
-            foreach (var type in this.ConstraintTypesNoUseSiteDiagnostics)
+
+            if (this.HasUnmanagedTypeConstraint)
             {
-                switch (type.SpecialType)
-                {
-                    case SpecialType.System_Object:
-                        // Avoid emitting unnecessary object constraint.
-                        continue;
-                    case SpecialType.System_ValueType:
-                        seenValueType = true;
-                        break;
-                }
-                var typeRef = moduleBeingBuilt.Translate(type,
-                                                         syntaxNodeOpt: (CSharpSyntaxNode)context.SyntaxNodeOpt,
-                                                         diagnostics: context.Diagnostics);
+                var typeRef = moduleBeingBuilt.Translate(
+                    typeSymbol: moduleBeingBuilt.Compilation.GetSpecialType(SpecialType.System_ValueType),
```

Comment:
>moduleBeingBuilt.Compilation.GetSpecialType(SpecialType.System_ValueType) [](start = 32, length = 73)

Where do we check goodness of this type?

Reply:
Vlad changed that bit. Also we check this during binding: `BindTypeParameterConstraints`. We also have tests in case this type is missing.

---
In reply to: [172415526](https://github.com/dotnet/roslyn/pull/24817#discussion_r172415526) [](ancestors = 172415526,172414868)